numpy/core/src/npymath/npy_math_complex.c.src

```diff
@@ -445,28 +445,26 @@ npy_cpow@c@ (@ctype@ a, @ctype@ b)
     @type@ bi = npy_cimag@c@(b);
     @ctype@ r;
 
-    if (br == 0. && bi == 0.) {
-        return npy_cpack@c@(1., 0.);
-    }
-    if (ar == 0. && ai == 0.) {
-        if (br > 0 && bi == 0) {
-            return npy_cpack@c@(0., 0.);
+    /*Checking if in a^b, if b is zero.
+    If a is not zero then by definition of logarithm a^0 is one.
+    If a is also zero then as per IEEE 0^0 is best defined as 1.*/
+    if(br == 0. && bi == 0.){
+        return npy_cpack@c@(1.,0.);
+    }
+    /*Here it is already tested that for a^b, b is not zero.
+    So we are checking behaviour of a.*/
+    else if(ar == 0. && ai == 0.){
+        /*Checking if real part of power is greater than zero then the result is zero.
+        If not the result is undefined as it blows up or oscillates.*/
+        if(br > 0){
+            return npy_cpack@c@(0.,0.);
         }
-        else {
-            volatile @type@ tmp = NPY_INFINITY@C@;
-            /*
-             * NB: there are four complex zeros; c0 = (+-0, +-0), so that
-             * unlike for reals, c0**p, with `p` negative is in general
-             * ill-defined.
-             *
-             *     c0**z with z complex is also ill-defined.
-             */
-            r = npy_cpack@c@(NPY_NAN@C@, NPY_NAN@C@);
-
-            /* Raise invalid */
-            tmp -= NPY_INFINITY@C@;
-            ar = tmp;
-            return r;
+        else{
+            //Generating an invalid
+            volatile @type@ tmp=NPY_INFINITY@C@;
+            tmp-=NPY_INFINITY@C@;
+            ar=tmp;
```

Comment:
Please fix the spaces here again, no operators without spaces around them.

Reply:
@seberg Here is what I found.  
The first section discusses how the complex exponentiation is done.  
Second section deals with the sign.  

**Complex Exponentiation**
If z and w are two complex numbers then  
z^w = e^{w*(ln|z|+i\*arg(z} \*e^{i\*w\*(2\*n\*pi)}
which means that z^w is multivalued. We call the value at ```n=0``` the principle value.  
Now, here I will discuss the case of 0^{non-zero complex number} which is related to our issue.  
0^w = e^{w \* (ln(0) + i\*arg(0,0))}\*e^{ i\*w\*(2\*n *pi)}  
Lets call this equation 1.  
Now arg is calculated with the help of atan2. Mathematically undefined but is realized in computer languages.
As we are considering C language, for unsigned 0, we get the output as 0. For signed numbers, we get the
output 0, pi, -0, -pi. This is demonstrated with the following C program.  

```C
#include<stdio.h>
#include<math.h>

int main(){
    double a=atan2(0,0);
    double b=atan2(+0.0,+0.0);
    double c=atan2(+0.0,-0.0);
    double d=atan2(-0.0,+0.0);
    double e=atan2(-0.0,-0.0);
    printf("%lf\n%lf\n%lf\n%lf\n%lf\n",a,b,c,d,e);
    return 0;
}
```

Output  
```
0.000000
0.000000
3.141593
-0.000000
-3.141593
```
Therefore e^{i\*arg(0,0)} will be +1 or -1 depending upon values of ar and ai (0 but still signed).  
Now, let's spilt equation 1 into three parts.  
Part 1: e^{ w\*ln|0| } (ln|0| = -inf)  
Part 2: e^{ i\*arg(0,0) }= ±1  
Part 3: e^{ i\*w*(2n*pi)}  
For complex number w , we get 4 cases
br = 0, bi != 0  
br != 0, bi = 0
br = 0, bi != 0
br != 0, bi != 0  
The case for br=0 and bi=0 leads to 0^0 which is defined to be 1 by IEEE.  
For case br != 0, bi = 0, we have
```
    if br > 0 we have
        Part 1= e^{ br* -inf }= 0
        Part2 = ±1 (actual sign depending on signs of ar and ai)
        Part3 = dont care 
        Result = ±0
    if br < 0  we have 
        Part 1= e^{ br*-inf }= inf
        Part 2= ±1 
        Part 3= oscillates between positive and negative value
        Therefore this leads to indeterminate form of inf*0 and
        the result is undefined . (This might also expalin why C 
        generates (inf,nan) as its output when encountered with 
        such values. In our case we generate (nan,nan).
```
For br = 0, bi != 0  
```
    if bi > 0
        Part 1: e^{ i*bi*-inf}
        which can be written as
        cos(bi*-inf)+isin(bi*-inf)
        which is undefined.
        Part 2: don't care
        Part 3: don't care
        Result: undefined
    if bi < 0
        Same as above because this time we get
        +inf instead of minus inf.
```
For br !=0 and bi != 0  
```
    Part 1: e^{ w * -inf}
    =e^{br*-inf} * e^{i*bi*-inf} (w=br + i*bi)
    Therefore if br > 0 this goes down to 0
    if not, the result oscillates and is undifined.
    Part 2 and 3 are dont care.
```
All the cases are summarized in the table below  
|      | bi < 0 | bi = 0 | bi > 0 |
|------|--------|--------|--------|
|br < 0|  undef |  undef |  undef |
|br = 0|  undef |    1   |  undef |
|br > 0|    0   |    0   |    0   |  

        Value of 0 ^ {complex}
Note: 0s are signed.

Combining with the complex exponentiation table  
|                   | exp = 0 |  exp = non-zero |
|-------------------|---------|-----------------|
|mantissa = 0       |    1    |    Table above  |
|mantissa = non-zero|    1    | Already handled |

**As we learned that complex exponentiation is multi-valued and the actual**
**signs depend on many factors such as signs of ar and ai and whether the API**
**is selecting principle value or not. Also 0+0i, 0-0i,-0+0i and -0-0i correspond**
**to the same point on the complex plane. So as long the result is 0 it mathematically**
**does not affect the result**  

I am ready to go with whatever you suggest. If you suggest going with signs (the resulting 0 contains sign opposite of bi as the C program below shows.)
I will make changes in the code accordingly. If you say its fine just make the code according to numpy C style I will do that.
```C
#include<stdio.h>
#include<math.h>
#include<complex.h>

int main(){
    complex mantissa[]={ 0.0+0.0*I, 0.0-0.0*I, -0.0+0.0*I, -0.0-0.0*I};
    complex exponents[]={1.0+1.0*I, 1.0-1.0*I, -1.0+1.0*I, -1.0-1.0*I};
    for(int i=0;i<4;i++){
        for(int j=0;j<4;j++){
            complex c=cpow(mantissa[i],exponents[j]);
            printf("%lf + %lf *I\n",creal(c),cimag(c));
        }
    }
    return 0;
}
```
Output:
```
0.000000 + -0.000000 *I
0.000000 + 0.000000 *I
inf + -nan *I
inf + -nan *I
0.000000 + -0.000000 *I
0.000000 + 0.000000 *I
inf + -nan *I
inf + -nan *I
0.000000 + -0.000000 *I
0.000000 + 0.000000 *I
inf + -nan *I
inf + -nan *I
0.000000 + -0.000000 *I
0.000000 + 0.000000 *I
inf + -nan *I
inf + -nan *I
```
**References:**
[Complex Eponentiation](https://brilliant.org/wiki/complex-exponentiation/#:~:text=Complex%20exponentiation%20extends%20the%20notion,y%20is%20a%20complex%20number.)  
[atan2](https://en.wikipedia.org/wiki/Atan2#Realizations_of_the_function_in_common_computer_languages)

Reply:
Sorry for taking a while to respond.  That anlysis sounds great, and `atan2` is a seems like a very good "basis" to make a decision of here I think, thanks!  (I have not followed every step, I admit)

My feeling is that we should try to preserve the signs for the 0 return.  And the fact that GCC gives a floating point warnings seems like a small issue in glibc?  After all, `atan2` makes a point of getting this right as well.  Also the old code-comment to me reads like: Let's return NaN, because the sign means that returning 0 would not be quite right.

So my personal judgment would be: return the correct sign here, in the special case!  And IIRC the rest already looked good.  I double checked the C99 standard, but could not find anything relevant (the `cpow` documentation is pretty much a stub).

I think we should open an issue on gcc/glibc about the warning flags being set for those signed 0 returns.  In the best case scenario we learn something new. In the other best case scenario glibc stops giving spurious warnings ;).

Reply:
Ah GLIBC issue of course, I have opened a bug report, hopefully in the right place here: https://sourceware.org/bugzilla/show_bug.cgi?id=27875